executor/mppcoordmanager/mpp_coordinator_manager.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mppcoordmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/mpp"
+	"github.com/pingcap/tidb/executor/metrics"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/copr"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// InstanceMPPCoordinatorManager is a local instance mpp coordinator manager
+var InstanceMPPCoordinatorManager = newMPPCoordinatorManger()
+
+const (
+	detectFrequency = 5 * time.Minute
+)
+
+// CoordinatorUniqueID identifies a unique coordinator
+type CoordinatorUniqueID struct {
+	MPPQueryID kv.MPPQueryID
+	GatherID   uint64
+}
+
+// MPPCoordinatorManager manages all mpp coordinator instances
+type MPPCoordinatorManager struct {
+	mu             sync.Mutex
+	serverOn       bool
+	serverAddr     string // empty if server is off
+	coordinatorMap map[CoordinatorUniqueID]kv.MppCoordinator
+	wg             sync.WaitGroup
+	ctx            context.Context
+	cancel         context.CancelFunc
+	maxLifeTime    uint64 // in Nano
+}
+
+// Run use a loop to detect and remove out of time Coordinators
+func (m *MPPCoordinatorManager) Run() {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	m.wg.Add(1)
+	m.maxLifeTime = uint64(copr.TiFlashReadTimeoutUltraLong.Nanoseconds() + detectFrequency.Nanoseconds())
+	go func() {
+		defer m.wg.Done()
+		ticker := time.NewTicker(detectFrequency)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-m.ctx.Done():
+				return
+			case <-ticker.C:
+				m.detectAndDelete(uint64(time.Now().UnixNano()))
+			}
+		}
+	}()
+}
+
+func (m *MPPCoordinatorManager) detectAndDelete(nowTs uint64) {
+	var outOfTimeIDs []CoordinatorUniqueID
+	m.mu.Lock()
+	for id := range m.coordinatorMap {
+		if nowTs > id.MPPQueryID.QueryTs+m.maxLifeTime {
+			outOfTimeIDs = append(outOfTimeIDs, id)
+			delete(m.coordinatorMap, id)
```

Comment:
Is it necessary to actively trigger the failure of the coodinator here, and terminate the mpp query?

Reply:
This background routine is added to ensure mpp coordinator map won't keep terminated coordinators, won't affect coordinator's life cycle. Even if coordinator was deleted from this map before it terminated, no critical error will happen.

Reply:
ok